File: old/app/excel/ReportingController.scala
package excel

import java.time.LocalDate
import java.time.format.DateTimeFormatter

import frontend.AuthenticatedUserAction
import javax.inject.{Inject, Singleton}
import play.api.libs.json.{Json, OWrites}
import play.api.mvc._
import reporting.ReportingModel.{Project, TimeEntry, ViewModel}
import reporting.TransformTogglToReporting
import toggl.TogglClient

import scala.concurrent.ExecutionContext

@Singleton
class ReportingController @Inject()(cc: ControllerComponents,
                                    WithBasicAuth: AuthenticatedUserAction,
                                    c: TogglClient,
                                    t: TransformTogglToReporting,
                                    e: TimesheetService)(implicit ec: ExecutionContext)
  extends AbstractController(cc) {

  def entries(client: Long, from: Option[LocalDate], to: Option[LocalDate], tagged: Option[Boolean]) = WithBasicAuth.async { implicit request =>
    implicit val projectWrites: OWrites[Project] = Json.writes[Project]
    implicit val timeEntryWrites: OWrites[TimeEntry] = Json.writes[TimeEntry]
    implicit val jsonWrites: OWrites[ViewModel] = Json.writes[ViewModel]

    val definiteTo = to.getOrElse(LocalDate.now().plusMonths(1).withDayOfMonth(1).minusDays(1))
    val definiteFrom = from.getOrElse(definiteTo.withDayOfMonth(1))

    for {
      entries <- c.entries(clientId = client, from = definiteFrom, to = definiteTo, nonBilledOnly = tagged.getOrElse(false))
      t <- t.transformInput(entries, definiteFrom, definiteTo, client)
    } yield Ok(Json.toJsObject(t))
  }

  def timesheet(client: Long, from: LocalDate, to: LocalDate) = WithBasicAuth.async { implicit request =>
    for {
      entries <- c.entries(clientId = client, from = from, to = to, nonBilledOnly = true)
      transformed <- t.transformInput(entries, from, to, client)
      excelModel <- e.transformViewModelToTimesheetModel(transformed)
      excel <- e.generateExcel(excelModel)
      datum = entries.data.headOption.map(_.start.toLocalDate).getOrElse(LocalDate.now())
    } yield Ok(excel)
      .as("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet")
      .withHeaders(CONTENT_DISPOSITION -> s"attachment; filename=${excelModel.name}-${datum.format(DateTimeFormatter.ofPattern("MM-yyyy"))}.xlsx")
  }
}

File: server-mn/src/main/kotlin/io/github/titaniumcoder/toggl/reporting/reporting/ReportingController.kt
package io.github.titaniumcoder.toggl.reporting.reporting

import io.micronaut.core.convert.format.Format
import io.micronaut.http.HttpHeaders
import io.micronaut.http.HttpResponse
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Get
import io.micronaut.http.annotation.PathVariable
import io.micronaut.http.annotation.QueryValue
import io.micronaut.security.annotation.Secured
import java.time.LocalDate
import java.time.format.DateTimeFormatter

@Controller("/api")
class ReportingController(val service: ReportingService) {
    @Secured("isAuthenticated()")
    @Get("/client/{clientId}")
    fun entries(
            @PathVariable("clientId") clientId: Long,
            @QueryValue("from") @Format("yyyy-MM-dd") from: LocalDate?,
            @QueryValue("to") @Format("yyyy-MM-dd") to: LocalDate?
    ) =
            service.entries(clientId, from, to)

    @Secured("isAuthenticated()")
    @Get("/timesheet/{clientId}", produces = ["application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"])
    fun timesheet(
            @PathVariable("clientId") clientId: Long,
            @QueryValue("from") @Format("yyyy-MM-dd") from: LocalDate,
            @QueryValue("to") @Format("yyyy-MM-dd") to: LocalDate
    ): HttpResponse<ByteArray> {
        val sheet = service.timesheet(clientId, from, to)

        val filename = "${sheet.name.toUpperCase()}-${sheet.date.format(DateTimeFormatter.ofPattern("yyyy-MM"))}.xlsx"

        return HttpResponse
                .ok(sheet.excel)
                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"$filename\"")
    }
}
